[New Rule] Potential Malicious File Downloaded from Google Drive

terrancedejesus opened the pull request:
## Issues
* https://github.com/elastic/security-team/issues/6719#event-9321167658

## Summary
Identifies potential malicious file download and execution from Google Drive. The rule checks for download activity from
Google Drive URL, followed by the creation of files commonly leveraged by or for malware. This could indicate an attempt
to run malicious scripts, executables or payloads.

Additional Information:
To emulate this activity, a shellscript `hello_world.sh` and PE `hello_world.exe` were created and hosted on a separate Google Drive. A python script, shellscript and PowerShell script were then used to download the files and execute them on the endpoints by using the URL `https://drive.google.com/uc?export=download&confirm=no_antivirus&id=${file_id}`. It is uncommon to download a file this way programmatically as backend developers would use OAuth GCP credentials to access the drive and download the file. This also indicates that the file is publicly available. The `confirm=no_antivirus` parameter in the URI tells Google to ignore that this may be a malicious file or else an HTML page is returned indicating so, again making this anomalous.

This rule is meant to be cross-platform as it could work for macOS, Linux and Windows.

<img width="2557" alt="Screenshot 2023-06-19 at 1 15 09 PM" src="https://github.com/elastic/detection-rules/assets/99630311/fb5822eb-0297-4e58-bec2-d74764ed7533">

#### VT Search Results for Requests to Google URL
<img width="2026" alt="Screenshot 2023-06-21 at 3 12 33 PM" src="https://github.com/elastic/detection-rules/assets/99630311/76d4fa92-faf3-423a-9385-5c775eae3e40">

terrancedejesus commented on rules/cross-platform/command_and_control_google_drive_malicious_file_download.toml:
```diff
@@ -0,0 +1,87 @@
+    "Threat Detection",
+    "Execution",
+    "Command and Control",
+    "Initial Access",
```
```suggestion
```

Aegrah commented on rules/cross-platform/command_and_control_google_drive_malicious_file_download.toml:
```diff
@@ -0,0 +1,87 @@
+
+    /* To remove FPs, only flag if these DNS requests or connection attempts are from common LoLBins */
+    process.name: (
+                    "bitsadmin.exe", "certutil.exe", "esentutl.exe", "wmic.exe", "PowerShell.exe",
```
Any specific reason cmd.exe is not here?

terrancedejesus replied:
Added, thanks @Aegrah!

Aegrah commented on rules/cross-platform/command_and_control_google_drive_malicious_file_download.toml:
```diff
@@ -0,0 +1,87 @@
+                    "curl", "wget", "scp", "ftp", "python", "perl", "ruby"))]
+
+/* Identify the creation of files following Google Drive connection with extensions commonly used for executables or libraries */
+[file where event.action == "creation" and file.extension: ("exe", "dll", "scr", "jar", "pif", "app", "dmg", "pkg", "so", "bin", "deb", "rpm","sh","hta","lnk")]
```
```suggestion
[file where event.action == "creation" and file.extension: ("exe", "dll", "scr", "jar", "pif", "app", "dmg", "pkg", "elf", "so", "bin", "deb", "rpm","sh","hta","lnk")]
```
```suggestion
[file where event.action == "creation" and file.extension: ("exe", "dll", "scr", "jar", "pif", "app", "dmg", "pkg", "so", "bin", "deb", "rpm","sh","hta","lnk")]
```
```suggestion
[file where event.action == "creation" and file.extension: ("exe", "dll", "scr", "jar", "pif", "app", "dmg", "pkg", "elf", "so", "bin", "deb", "rpm","sh","hta","lnk")]
```

Aegrah commented on rules/cross-platform/command_and_control_google_drive_malicious_file_download.toml:
```diff
@@ -0,0 +1,87 @@
+                    "curl", "wget", "scp", "ftp", "python", "perl", "ruby"))]
+
+/* Identify the creation of files following Google Drive connection with extensions commonly used for executables or libraries */
+[file where event.action == "creation" and file.extension: ("exe", "dll", "scr", "jar", "pif", "app", "dmg", "pkg", "so", "bin", "deb", "rpm","sh","hta","lnk")]
```
```suggestion
[file where event.action == "creation" and file.extension: ("exe", "dll", "scr", "jar", "pif", "app", "dmg", "pkg", "elf", "so", "bin", "deb", "rpm","sh","hta","lnk")]
```

terrancedejesus replied:
ooo good one, those pesky ELF binaries 😄

Aegrah commented on rules/cross-platform/command_and_control_google_drive_malicious_file_download.toml:
```diff
@@ -0,0 +1,87 @@
+    dns.question.name == "drive.google.com" and dns.question.type == "A") or
+
+    /* Look for connection attempts to address that resolves to Google */
+    (destination.as.organization.name == "GOOGLE" and event.action == "connection_attempted" and
```
I am wondering, would it be possible to only look at the `dns.question.name == "drive.google.com" and dns.question.type == "A")` and then be less strict in using LolBins? Because the `(destination.as.organization.name == "GOOGLE" and event.action == "connection_attempted"` is going to occur very frequently, especially for VMs in GCP. 

In the samples I detonated I noticed that many binaries tried to connect to a Google AS without using any LoLbins, just litteraly straight from the binary that was downloaded. As this rule specifically looks for potential malicious files downloaded from G-drive I think that could make sense, but if you want to detect all file downloads from maliciously hosted C2's in GCP as well I understand your current approach makes the most sense by far.

terrancedejesus replied:
Thanks for the additional thoughts. Unfortunately I was hoping that the Defend integration and endpoint network data would have more application layer visibility, but it did not seem so. Therefore I had to rely on DNS requests, the only network data available when doing this or else I would be focused solely on the HTTP requests.

Your right, the connection attempted to an address that resolves to Google is not what I had hoped for, but grouped together with the LolBins should help reduce that noise. When testing the query, the data I had generated was from a Windows and Linux VM in GCP and this filtered out those additional connection attempts. Since this is an `OR` statement, the DNS request or the direct Google connection attempt will work for the second sequence so it shouldn't miss any random binaries doing this.

brokensound77 commented on rules/cross-platform/command_and_control_google_drive_malicious_file_download.toml:
```diff
@@ -0,0 +1,86 @@
+type = "eql"
+query = '''
+sequence by host.id, user.id with maxspan=30s
+[any where (
```
#2811 will allow you to set the category field, to avoid using `any`

terrancedejesus replied:
Awesome! Yea the only reason it is `any` is because I allow the `event.action` to be load which is from a library event category while keeping `process` related values such as `start` in the first sequence as well.

brokensound77 commented on rules/cross-platform/command_and_control_google_drive_malicious_file_download.toml:
```diff
@@ -0,0 +1,86 @@
+language = "eql"
+license = "Elastic License v2"
+name = "Potential Malicious File Downloaded from Google Drive"
+note = """## Setup
```
we can probably remove this now that min stack > 8.2 (outside of the scope of this PR)

terrancedejesus replied:
Removed, thanks for the reminder!

Aegrah reviewed (approved):
LGTM!

brokensound77 commented on rules/cross-platform/command_and_control_google_drive_malicious_file_download.toml:
```diff
@@ -0,0 +1,82 @@
+    /* Look for processes started on Linux and macOS */
+    event.action: ("exec","fork"))]
+
+[network where (
+    /* Look for DNS requests for Google Drive */
+    dns.question.name == "drive.google.com" and dns.question.type == "A") or
```
nit

```suggestion
[network where 
    /* Look for DNS requests for Google Drive */
    (dns.question.name == "drive.google.com" and dns.question.type == "A") or
```

brokensound77 commented on rules/cross-platform/command_and_control_google_drive_malicious_file_download.toml:
```diff
@@ -0,0 +1,82 @@
+    dns.question.name == "drive.google.com" and dns.question.type == "A") or
+
+    /* Look for connection attempts to address that resolves to Google */
+    (destination.as.organization.name == "GOOGLE" and event.action == "connection_attempted" and
```
same question for `destination.as.organization.name` using `:` vs `==`

brokensound77 commented on rules/cross-platform/command_and_control_google_drive_malicious_file_download.toml:
```diff
@@ -0,0 +1,82 @@
+    (destination.as.organization.name == "GOOGLE" and event.action == "connection_attempted" and
+
+    /* To remove FPs, only flag if these DNS requests or connection attempts are from common LoLBins */
+    process.name: (
```
++ great choice to use inclusion by exception!

terrancedejesus replied:
I added a comment with this filter for future reference if needed. If we add these, then we may miss malware that do not rely on LOLBins but rather have this capability built in.

Let's cast the net wider for now with potential tuning after telemetry review. Thoughts @brokensound77 ?

terrancedejesus commented:
#### Update
- Removed requirement for LOLBins in the second sequence with network related traffic. Reason - This would blind us to binaries that are not LOLBin making direct connection attempts or DNS lookups for Google Drive. I have kept this as a comment for potential future tuning.
- Removed `code_signature` value requirements in the first sequence. This is based on testing and because eliminate there are two alternatives here. The first being a malicious binary starting that does not have a code signature or it is not trusted, however, for instance, a malicious PowerShell script could be ran where the code signature reflects Microsoft and is trusted since `powershell.exe` would be spawned to handle this.
- Changed the sequence to be `host.id` and `process.entity_id` as this narrows the scope to a single host where the `entity_id` is consistent.

@brokensound77 @Samirbous - Any additional thoughts on these changes?

terrancedejesus commented on rules/cross-platform/command_and_control_google_drive_malicious_file_download.toml:
```diff
@@ -0,0 +1,77 @@
+sequence by host.id, process.entity_id with maxspan=30s
+[any where
+
+    /* Look for processes started or libraries loaded from untrusted or unsigned Windows and macOS binaries */
```
```suggestion
    /* Look for processes started or libraries loaded from untrusted or unsigned Windows, Linux or macOS binaries */
```

brokensound77 commented on rules/cross-platform/command_and_control_google_drive_malicious_file_download.toml:
```diff
@@ -0,0 +1,77 @@
+]
+
+/* Identify the creation of files following Google Drive connection with extensions commonly used for executables or libraries */
+[file where event.action == "creation" and file.extension: (
```
```suggestion
[file where event.action == "creation" and file.extension : (
```

brokensound77 commented:
>Any additional thoughts on these changes?

good to see `process.entity_id` swapped in, which will eliminate a lot of potential FP. The rest makes sense to me

brokensound77 reviewed (approved):
Thanks for the changes, LGTM 🚀
